java/arcs/core/allocator/Allocator.kt

```diff
@@ -107,26 +109,40 @@ class Allocator(val hostRegistry: HostRegistry) {
      * Finds [HandleConnection] instances which were unresolved at build time
      * [CreateableStorageKey]) and attaches generated keys via [CapabilitiesResolver].
      */
-    private fun createStorageKeysIfNecessary(arcId: ArcId, idGenerator: Id.Generator, plan: Plan) {
+    private fun createStorageKeysIfNecessary(
+        arcId: ArcId,
+        idGenerator: Id.Generator,
+        plan: Plan
+    ): Plan {
         val createdKeys: MutableMap<StorageKey, StorageKey> = mutableMapOf()
+        val allHandles = Plan.particleLens.traverse() + Plan.Particle.handlesLens.traverse()
+
+        return allHandles.mod(plan) { handle ->
+            Plan.HandleConnection.storageKeyLens.mod(handle) {
+                replaceCreateKey(createdKeys, arcId, idGenerator, it, handle.type)
+            }
+        }
+    }
 
-        plan.particles.forEach {
-            it.handles.values.forEach { spec ->
-                spec.apply {
-                    if (storageKey is CreateableStorageKey) {
-                        if (!createdKeys.containsKey(storageKey)) {
-                            createdKeys[storageKey] = createStorageKey(
-                                arcId,
-                                idGenerator,
-                                storageKey as CreateableStorageKey,
-                                type
-                            )
-                        }
-                        storageKey = createdKeys[storageKey]!!
-                    }
-                }
+    fun replaceCreateKey(
+        createdKeys: MutableMap<StorageKey, StorageKey>,
+        arcId: ArcId,
+        idGenerator: Id.Generator,
+        storageKey: StorageKey,
+        type: Type
+    ): StorageKey {
+        if (storageKey is CreateableStorageKey) {
+            if (!createdKeys.containsKey(storageKey)) {
+                createdKeys[storageKey] = createStorageKey(
+                    arcId,
+                    idGenerator,
+                    storageKey,
+                    type
+                )
             }
+            return createdKeys[storageKey]!!
```

Comment:
I believe you could use `getOrPut` here to avoid the null check.

Reply:
nice